bam_consensus.c

```diff
@@ -0,0 +1,1426 @@
+//        Specifically het calls where mods are predominantly on one
+//        strand.  So maybe require + and - calls and check concordance
+//        before calling a het as confident.  (Still call, but low qual?)
+
+// Eg 50T 20A seems T/A het,
+// but 30T+ 20T- 18A+ 2A- seems like a consistent A miscall on one strand
+// only, while T is spread evenly across both strands.
+
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <limits.h>
+#include <float.h>
+#include <ctype.h>
+
+#include <htslib/sam.h>
+
+#include "samtools.h"
+#include "sam_opts.h"
+#include "bam_plbuf.h"
+
+#ifndef MIN
+#  define MIN(a,b) ((a)<(b)?(a):(b))
+#endif
+#ifndef MAX
+#  define MAX(a,b) ((a)>(b)?(a):(b))
+#endif
+
+// Minimum cutoff for storing mod data; => at least 10% chance
+#define MOD_CUTOFF 0.46
+
+enum format {
+    FASTQ,
+    FASTA,
+    PILEUP
+};
+
+typedef unsigned char uc;
+
+typedef struct {
+    // User options
+    char *reg;
+    int use_qual;
+    int use_mqual;
+    int min_mqual;
+    int max_mqual;
+    int min_depth;
+    double call_fract;
+    double het_fract;
+    int gap5;
+    enum format fmt;
+    int cons_cutoff;
+    int ambig;
+    int line_len;
+    int default_qual;
+    int het_only;
+    int all_bases;
+    int show_del;
+    int show_ins;
+
+    // Internal state
+    samFile *fp;
+    FILE *fp_out;
+    sam_hdr_t *h;
+    hts_idx_t *idx;
+    hts_itr_t *iter;
+    kstring_t ks_line;
+    kstring_t ks_ins_seq;
+    kstring_t ks_ins_qual;
+} consensus_opts;
+
+/*
+ * Reads a single alignment record, using either the iterator
+ * or a direct sam_read1 call.
+ */
+static int readaln(void *data, bam1_t *b) {
+    consensus_opts *dat = (consensus_opts *)data;
+    if (dat->iter)
+        return sam_itr_next(dat->fp, dat->iter, b);
+    else
+        return sam_read1(dat->fp, dat->h, b);
+}
+
+static const char *seq_nt16_str_pad = "=ACMGRSVTWYHKDBN=acmgrsvtwyhkdbn";
+
+/*
+ * Produce the insertion portion of the consensus sequence.
+ * Returns 0 on success, filling out ins_ks and ins_ks,
+ *        -1 on failure
+ */
+static int insertion_consensus(const bam_pileup1_t *plp, int nplp,
+                               consensus_opts *opts, int nins, int ins_len,
+                               kstring_t *ins_ks, kstring_t *inq_ks,
+                               kstring_t *tmp_ks) {
+    // Map "seqi" nt16 to A,C,G,T compatibility with weights on pure bases.
+    // where seqi is A | (C<<1) | (G<<2) | (T<<3)
+    //                        * A C M  G R S V  T W Y H  K D B N
+    static int seqi2A[16] = { 0,8,0,4, 0,4,0,2, 0,4,0,2, 0,2,0,1 };
+    static int seqi2C[16] = { 0,0,8,4, 0,0,4,2, 0,0,4,2, 0,0,2,1 };
+    static int seqi2G[16] = { 0,0,0,0, 8,4,4,1, 0,0,0,0, 4,2,2,1 };
+    static int seqi2T[16] = { 0,0,0,0, 0,0,0,0, 8,4,4,2, 8,2,2,1 };
+
+    int ins_[128][16] = {0}, (*ins)[16] = ins_;
+    int i;
+
+    if (ins_len > 128)
+        if (!(ins = calloc(ins_len * 16, sizeof(int))))
```

Comment:
This is quite a big multiplier.  On 32-bit platforms you could get an overflow with a carefully-crafted insertion.  It would be better to do `calloc(ins_len, 16 * sizeof(int))`.